python/interpret_text/common/utils_introspective_rationale.py

```diff
@@ -0,0 +1,404 @@
+        self.target_sparsity = 0.3
+
+        # this is the target number of target continuous pieces
+        # it has no effect now, because lambda_continuity is 0.
+        self.count_pieces = 4
+
+        # rate at which the generator explores different rationales
+        self.exploration_rate = 0.05
+
+        # multiplier to reward/penalize an accuracy gap between the classifier
+        # and anti-classifier
+        self.lambda_acc_gap = 1.2
+
+        # learning rate
+        self.lr = 2e-4
+
+        # whether to tune the weights of the embedding layer in the RNN classifier module
+        # many of these weights are from passed in glove embeddings, but
+        # it could be useful to fine tune those weights that did not inherit glove vector values
+        self.fine_tuning = True
+
+        # training parameters
+        self.cuda = cuda
+        self.pretrain_cls = pretrain_cls
+        self.num_pretrain_epochs = num_pretrain_epochs
+        self.num_epochs = num_epochs
+        self.train_batch_size = batch_size
+        self.test_batch_size = batch_size
+
+        # stop training if validation acc does not improve for more than
+        # training_stop thresh
+        self.training_stop_thresh = 5
+
+        # the numerical labels for classification
+        # ex: MNLI needs [0, 1, 2, 3, 4]
+        self.labels = [0, 1]
+
+        # for saving models and logging
+        self.save_best_model = save_best_model
+        self.model_prefix = model_prefix
+        self.save_path = model_save_dir
+
+        self.model_folder_path = os.path.join(
+            self.save_path,
+            self.model_prefix + "_training_")
+
+        if not os.path.exists(self.model_folder_path):
+            os.makedirs(self.model_folder_path)
+        self.log_filepath = os.path.join(
+            self.model_folder_path, "training_stats.txt"
+        )
+        logging.basicConfig(
+            filename=self.log_filepath, filemode="a", level=logging.INFO
+        )
+
+
+class BertPreprocessor:
+    """A class that tokenizes and otherwise processes text to be encoded
+       by a BERT model.
+    """
+
+    def __init__(self, build_word_dict=False, tokenizer=None, max_length=50,
+                 pad_to_max=True, text=None):
+        """Initialize the BertPreprocessor.
+
+        :param tokenizer: an initialized tokenizer with 'encode_plus' and
+            'convert ids to tokens' methods; defaults to None
+        :type tokenizer: tokenizer, optional
+        :param max_length: the max number of tokens allowable in an example,
+            defaults to 50
+        :type max_length: int, optional
+        :param pad_to_max: whether to pad all examples to the same (max)
+            length, defaults to True
+        :type pad_to_max: bool, optional
+        """
+        if tokenizer:
+            self.tokenizer = tokenizer
+        else:
+            self.tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
```

Comment:
I think we don't save users enough code to warrant this dependency yet

It could be worth replacing:
self.tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")

with:
raise ValueError( tokenizer should be set, a commonly used one is transformers.BertTokenizer.from_pretrained("bert-base-uncased") or a link to their website)

Then we can remove the transformers dependency. Users may have spacy for tokenizing, we should not require both

Reply:
Worst case we can have a func get_tranformers_tokenizer() that raises with a pip install transformers if we want to improve usability